cmd/sysinfo-xl.go

```diff
@@ -0,0 +1,235 @@
+	var errs = make([]error, len(storageDisks))
+
+	var wg = &sync.WaitGroup{}
+
+	// Write `format.json` to all disks.
+	for index, disk := range storageDisks {
+		if disk == nil {
+			errs[index] = errDiskNotFound
+			continue
+		}
+		wg.Add(1)
+		go func(index int, disk StorageAPI, sysinfo *sysInfoConfig) {
+			defer wg.Done()
+			errs[index] = saveSysInfoXL(disk, sysinfo)
+		}(index, disk, sysinfo)
+	}
+
+	// Wait for the routines to finish.
+	wg.Wait()
+
+	writeQuorum := len(storageDisks)/2 + 1
+	return reduceWriteQuorumErrs(ctx, errs, nil, writeQuorum)
+}
+
+// initSysInfoXL - save XL format configuration on all disks.
+func initSysInfoXL(ctx context.Context, storageDisks []StorageAPI, setCount, disksPerSet int) (*sysInfoConfig, error) {
+	sysinfo := newSysInfo(len(globalAdminPeers))
+
+	// get the sysinfo from individual nodes
+	// and save it in this structure
+	var wg sync.WaitGroup
+	reply := make(map[xnet.Host]*ServerSystemInfo, len(globalEndpoints))
+	for _, ep := range globalEndpoints {
+		if ep.IsLocal {
+			var err error
+			sysinfo.NodesSysInfo[0], err = localServerSystemInfo()
+			if err != nil {
+				reqInfo := (&logger.ReqInfo{}).AppendTags("peerAddress", ep.String())
+				ctx := logger.SetReqInfo(context.Background(), reqInfo)
+				logger.LogIf(ctx, err)
+				return nil, err
+			}
+		}
+	}
+	peerRPCClientMap := makeRemoteRPCClients(globalEndpoints)
+	/*
+		// Gather server information for all nodes
+		for i, p := range globalAdminPeers {
+			wg.Add(1)
+
+			// Gather information from a peer in a goroutine
+			go func(idx int, peer adminPeer) {
+				defer wg.Done()
+
+				// Initialize server info at index
+				reply[idx] = ServerSystemInfo{Addr: peer.addr}
+
+				serverSystemInfoData, err := peer.cmdRunner.ServerSystemInfo()
+				if err != nil {
+					reqInfo := (&logger.ReqInfo{}).AppendTags("peerAddress", peer.addr)
+					ctx := logger.SetReqInfo(context.Background(), reqInfo)
+					logger.LogIf(ctx, err)
+					reply[idx].Error = err.Error()
+					return
+				}
+
+				reply[idx].Data = &serverSystemInfoData
+			}(i, p)
+		}*/
+	for k, v := range peerRPCClientMap {
+		wg.Add(1)
+
+		// Gather information from a peer in a goroutine
+		go func(idx xnet.Host, peer *PeerRPCClient) {
+			defer wg.Done()
+
+			// Initialize server info at index
+			reply[idx] = &ServerSystemInfo{Addr: idx.Name + strconv.Itoa(int(idx.Port))}
+
+			serverSystemInfoData, err := peer.GetSysInfo()
+			if err != nil {
+				reqInfo := (&logger.ReqInfo{}).AppendTags("peerAddress", idx.Name+strconv.Itoa(int(idx.Port)))
+				ctx := logger.SetReqInfo(context.Background(), reqInfo)
+				logger.LogIf(ctx, err)
+				reply[idx].Error = err.Error()
+				return
+			}
+
+			reply[idx].Data = &serverSystemInfoData
+		}(k, v)
+	}
+	wg.Wait()
+	i := 1
+	for _, sysinfodata := range reply {
+		if sysinfodata.Error != "" {
+			return nil, errors.New(sysinfodata.Error)
+		}
+		sysinfo.NodesSysInfo[i] = *sysinfodata.Data
+		i = i + 1
```

Comment:
sysinfo.NodesSysInfo is already initialized with len(globalAdminPeers) elements. Isn't this better ?

 ```go
+	for i, sysinfodata := range reply {
+		if sysinfodata.Error != "" {
+			return nil, errors.New(sysinfodata.Error)
+		}
+		sysinfo.NodesSysInfo[i] = *sysinfodata.Data
+  }
```

Reply:
`reply` is of type `map[xnet.Host]*ServerSystemInfo`, so `i` in your case will not be the index.